src/System.Reflection.Metadata.ILDasm/src/Decoder/ILDecoder.cs

```diff
@@ -0,0 +1,582 @@
+                        instruction = new ILInt64Instruction(opCode, ilReader.ReadInt64(), -1, expectedSize + 8);
+                        break;
+                    case OperandType.InlineR:
+                        instruction = new ILDoubleInstruction(opCode, ilReader.ReadDouble(), -1, expectedSize + 8);
+                        break;
+                    case OperandType.InlineSwitch:
+                        instruction = CreateSwitchInstruction(ref ilReader, expectedSize, ilOffset, opCode);
+                        break;
+                    case OperandType.ShortInlineBrTarget:
+                        instruction = new ILInt16BranchInstruction(opCode, ilReader.ReadSByte(), ilOffset, expectedSize + 1);
+                        break;
+                    case OperandType.InlineBrTarget:
+                        instruction = new ILBranchInstruction(opCode, ilReader.ReadInt32(), ilOffset, expectedSize + 4);
+                        break;
+                    case OperandType.ShortInlineI:
+                        instruction = new ILByteInstruction(opCode, ilReader.ReadByte(), -1, expectedSize + 1);
+                        break;
+                    case OperandType.ShortInlineR:
+                        instruction = new ILSingleInstruction(opCode, ilReader.ReadSingle(), -1, expectedSize + 4);
+                        break;
+                    case OperandType.InlineNone:
+                        instruction = new ILInstructionWithNoValue(opCode, expectedSize);
+                        break;
+                    case OperandType.ShortInlineVar:
+                        byte token = ilReader.ReadByte();
+                        instruction = new ILInt16VariableInstruction(opCode, GetVariableName(opCode, token, _methodDefinition), token, expectedSize + 1);
+                        break;
+                    case OperandType.InlineVar:
+                        shortOperand = ilReader.ReadUInt16();
+                        instruction = new ILVariableInstruction(opCode, GetVariableName(opCode, shortOperand, _methodDefinition), shortOperand, expectedSize + 2);
+                        break;
+                    case OperandType.InlineSig:
+                        intOperand = ilReader.ReadInt32();
+                        instruction = new ILStringInstruction(opCode, GetSignature(mdReader, intOperand, provider), intOperand, expectedSize + 4);
+                        break;
+                    default:
+                        break;
+                }
+                ilOffset += instruction.Size;
+                yield return instruction;
+            }
+        }
+
+        internal static ILLocal[] DecodeLocalSignature(MethodBodyBlock methodBody, MetadataReader mdReader, ILTypeProvider provider)
+        {
+            if (methodBody.LocalSignature.IsNil)
+            {
+                return new ILLocal[0];
+            }
+            var localTypes = SignatureDecoder.DecodeLocalSignature(methodBody.LocalSignature, provider);
```

Comment:
We should try to avoid ```var``` when the right-side of the assignment doesn't make it very clear what the type is.